### PR TITLE
Record the amount of free pages instead of hashmapFreeCount calculation

### PR DESCRIPTION
**Pull request by xing0821**
reduce the cost of etcd compaction with record the amount of free pages

image

### CODE REVIEW

File: freelist_hmap.go
Comment (ptabor) [inline freelist_hmap.go:141]: For "completeness", please reset the counter in the init() method.

Comment (xing0821) [reply freelist_hmap.go:141]: thx for quickly reply~  fix it

File: freelist_hmap.go
Comment (ptabor) [inline freelist_hmap.go:7]: As such changes can be very subtle, I would feel safer if we 
could test them/verify extensively with logic of this style: 

```
if env.get("BBOLT_VERIFY")='true' {
   if f.freePagesCount != f.slowFreePagesCount() {
     panic(...);
   }
}
```

Even better having a piece of assertion framework: 

```
func assertF(assertFunc func () bool, fmt string, args string) {
  if env.get("BBOLT_VERIFY")='true' {
     if !assertFunc() {
        panic(fmt, args);
     }
 } 
}
```

to write: 

```
  assert(func() { f.freePagesCount == f.slowFreePagesCount()}, "freePagesCount is out of sync with free pages map");
```

**Review by ptabor — CHANGES_REQUESTED** [review]

**Comment by ptabor** [comment]
Nice finding !. 
I assume it fixed any bbolt commit, and not only etcd compaction. 

Could you plot the performance charts on etcd before and after this change using: 
  https://github.com/etcd-io/etcd/tree/main/tools/rw-heatmaps

File: freelist_hmap.go
Comment (xing0821) [reply freelist_hmap.go:8]: agree, add an assert to verify it will be safer!
os.GetEnv() will do syscall every single time, so I just init it as a global variable

File: freelist_hmap.go
Comment (ptabor) [inline freelist_hmap.go:7]: This code now will be executed always (so no performance benefit). 

It's the reason I proposed wrapping it in a func: 
```
_assertVerify(func() { return int(f.freePagesCount) == f.hashmapFreeCountSlow() } )
```

Comment (xing0821) [reply freelist_hmap.go]: Sorry for that, fix it~

**Comment by xing0821** [comment]
> Nice finding !.
> I assume it fixed any bbolt commit, and not only etcd compaction.
> 
> Could you plot the performance charts on etcd before and after this change using:
> https://github.com/etcd-io/etcd/tree/main/tools/rw-heatmaps

sure, there are a lot of test so I will report performance charts until the test case finished

File: db.go
Comment (lilic) [inline db.go:44]: Besides env var would it be possible to configure this in another way?

**Review by lilic — COMMENTED** [review]
Thank you!

Running benchmark would be great.

File: db.go
Comment (xing0821) [reply db.go:44]: maybe we can add a parameter in Options and init the 'var assertVerify' when user call the Open method

**Comment by cenkalti** [comment]
I run the rw-heatmaps benchmark on 2 m5.2xlarge instances in parallel.

# Read results
record-free-pages png_read

# Write results
record-free-pages png_write

The results show improved performance except when R/W ratio is 4. It doesn't make sense for this single case. I'll do another run in bigger instances to make sure the results are consistent.

**Comment by cenkalti** [comment]
I run rw-heatmaps benchmarks again, this time on 2 m5.16xlarge instances.

# Read results
record-free-pages-2 png_read


# Write results
record-free-pages-2 png_write

**Comment by ncabatoff** [comment]
I've encountered this being an issue in a large Vault installation that did a lot of deletions.  Compaction is the right thing to do, but sometimes it's hard to do in a timely manner without causing disruption.  The stats calculation this PR avoids is causing a notable service degradation in this installation, and our profiling shows the same thing as seen here.  @cenkalti you seem to be the maintainer who's most recently looked at the PR, is there anything further you'd like to see done before this lands?  Can I help?

**Comment by cenkalti** [comment]
In my opinion, this change is simple and safe. However, I am not allowed to merge PRs, I am only reviewer. You should reach out to @ahrtr .

File: freelist_hmap.go
Comment (ahrtr) [reply freelist_hmap.go:8]: Please follow 
- https://github.com/etcd-io/etcd/blob/main/client/pkg/verify/verify.go

To be clearer, pls add a verify.go under` internal/common`.  All verifications should be disabled in production, and enabled in test by default.

Comment (ncabatoff) [reply freelist_hmap.go:8]: @ahrtr, I fear that @xing0821 is no longer motivated to contribute to this PR.  Can you suggest a way to move forward?  I could open a new PR containing their changes, as well as the verify.go change you're suggesting, but I don't want them to lose credit for the work.

Comment (ahrtr) [reply freelist_hmap.go:8]: Thank you @ncabatoff .  I suggest to follow steps roughly below,
- fork a branch on top of this PR; so that @xing0821 is still the author of the first commit.
- rebase your PR;
- add your new change(s) as the new commits.

Does this make sense to you?

**Comment by ahrtr** [comment]
superseded by https://github.com/etcd-io/bbolt/pull/585

**Pull request closed without merging**
